File: sdk/agentlens/tracer.py
"""
AgentLens tracer: decorators that turn plain functions into DAG nodes.

    lens = AgentLens(endpoint="http://localhost:7430")

    @lens.trace("research_agent", max_cost_usd=0.10)
    def research_agent(query): ...

    @lens.span("retrieve_docs", kind=SpanKind.RETRIEVAL)
    def retrieve_docs(query): ...

Sync and async functions are supported identically. Export happens on a
background thread so tracing never blocks the agent.
"""

from __future__ import annotations

import functools
import inspect
import traceback
from typing import Any, Callable, Optional

from . import context as ctx
from .cost import estimate_cost_usd
from .exporters import Exporter, HttpExporter, ConsoleExporter
from .models import AgentRun, LLMMetadata, Span, SpanKind, SpanStatus, _preview


class BudgetExceeded(RuntimeError):
    """Raised when a run trips its token or cost budget guard."""

    def __init__(self, run: AgentRun, reason: str):
        super().__init__(f"AgentLens budget guard: {reason} (run={run.name})")
        self.run = run
        self.reason = reason


class AgentLens:
    """Client that owns configuration and produces tracing decorators."""

    def __init__(
        self,
        endpoint: Optional[str] = None,
        api_key: Optional[str] = None,
        exporter: Optional[Exporter] = None,
        cost_table: Optional[dict[str, tuple[float, float]]] = None,
        on_budget: str = "raise",  # "raise" | "pause" | "warn"
    ):
        if exporter is not None:
            self.exporter = exporter
        elif endpoint:
            self.exporter = HttpExporter(endpoint, api_key=api_key)
        else:
            self.exporter = ConsoleExporter()
        self.cost_table = cost_table or {}
        if on_budget not in ("raise", "pause", "warn"):
            raise ValueError("on_budget must be 'raise', 'pause', or 'warn'")
        self.on_budget = on_budget

    # ------------------------------------------------------------------ #
    # decorators
    # ------------------------------------------------------------------ #

    def trace(
        self,
        name: Optional[str] = None,
        tags: Optional[list[str]] = None,
        max_total_tokens: Optional[int] = None,
        max_cost_usd: Optional[float] = None,
        metadata: Optional[dict[str, Any]] = None,
    ) -> Callable:
        """Mark a function as an agent entrypoint. Creates an AgentRun."""

        def decorator(fn: Callable) -> Callable:
            run_name = name or fn.__name__

            def _start() -> tuple[AgentRun, Span, Any, Any]:
                run = AgentRun(
                    name=run_name,
                    tags=list(tags or []),
                    max_total_tokens=max_total_tokens,
                    max_cost_usd=max_cost_usd,
                    metadata=dict(metadata or {}),
                )
                root = Span(name=run_name, kind=SpanKind.AGENT)
                run.spans.append(root)
                run_token = ctx.set_run(run)
                span_token = ctx.set_span(root)
                return run, root, run_token, span_token

            def _finish(run, root, run_token, span_token, error: Optional[BaseException]):
                if error is None:
                    root.finish(SpanStatus.SUCCESS)
                    run.finish(SpanStatus.SUCCESS)
                elif isinstance(error, BudgetExceeded):
                    root.finish(SpanStatus.PAUSED, error=str(error))
                    run.finish(SpanStatus.PAUSED, error=str(error))
                else:
                    tb = "".join(traceback.format_exception(error)).strip()
                    root.finish(SpanStatus.ERROR, error=tb)
                    run.finish(SpanStatus.ERROR, error=str(error))
                ctx.reset_span(span_token)
                ctx.reset_run(run_token)
                try:
                    self.exporter.export(run)
                except Exception:
                    pass  # tracing must never take the agent down

            if inspect.iscoroutinefunction(fn):

                @functools.wraps(fn)
                async def async_wrapper(*args, **kwargs):
                    run, root, rt, st = _start()
                    root.inputs = _preview({"args": args, "kwargs": kwargs})
                    try:
                        result = await fn(*args, **kwargs)
                        root.outputs = _preview(result)
                        _finish(run, root, rt, st, None)
                        return result
                    except BaseException as e:
                        _finish(run, root, rt, st, e)
                        raise

                return async_wrapper

            @functools.wraps(fn)
            def wrapper(*args, **kwargs):
                run, root, rt, st = _start()
                root.inputs = _preview({"args": args, "kwargs": kwargs})
                try:
                    result = fn(*args, **kwargs)
                    root.outputs = _preview(result)
                    _finish(run, root, rt, st, None)
                    return result
                except BaseException as e:
                    _finish(run, root, rt, st, e)
                    raise

            return wrapper

        return decorator

    def span(
        self,
        name: Optional[str] = None,
        kind: SpanKind = SpanKind.CUSTOM,
        retries: int = 0,
    ) -> Callable:
        """Mark a function as a child step (DAG node) of the current run."""

        def decorator(fn: Callable) -> Callable:
            span_name = name or fn.__name__

            def _open(retry_of: Optional[str]) -> tuple[Optional[Span], Any]:
                run = ctx.current_run()
                if run is None:
                    return None, None  # untraced call: pass through
                parent = ctx.current_span()
                span = Span(
                    name=span_name,
                    kind=kind,
                    parent_id=parent.span_id if parent else None,
                    retry_of=retry_of,
                )
                run.spans.append(span)
                token = ctx.set_span(span)
                return span, token

            def _close(span: Span, token, error: Optional[BaseException]) -> None:
                if error is None:
                    span.finish(SpanStatus.SUCCESS)
                else:
                    tb = "".join(traceback.format_exception(error)).strip()
                    span.finish(SpanStatus.ERROR, error=tb)
                ctx.reset_span(token)
                self._check_budget()

            if inspect.iscoroutinefunction(fn):

                @functools.wraps(fn)
                async def async_wrapper(*args, **kwargs):
                    attempt, retry_of = 0, None
                    while True:
                        span, token = _open(retry_of)
                        if span is None:
                            return await fn(*args, **kwargs)
                        span.inputs = _preview({"args": args, "kwargs": kwargs})
                        try:
                            result = await fn(*args, **kwargs)
                            span.outputs = _preview(result)
                            _close(span, token, None)
                            return result
                        except BudgetExceeded:
                            _close(span, token, None)
                            raise
                        except BaseException as e:
                            _close(span, token, e)
                            if attempt < retries:
                                attempt += 1
                                retry_of = span.span_id
                                continue
                            raise

                return async_wrapper

            @functools.wraps(fn)
            def wrapper(*args, **kwargs):
                attempt, retry_of = 0, None
                while True:
                    span, token = _open(retry_of)
                    if span is None:
                        return fn(*args, **kwargs)
                    span.inputs = _preview({"args": args, "kwargs": kwargs})
                    try:
                        result = fn(*args, **kwargs)
                        span.outputs = _preview(result)
                        _close(span, token, None)
                        return result
                    except BudgetExceeded:
                        _close(span, token, None)
                        raise
                    except BaseException as e:
                        _close(span, token, e)
                        if attempt < retries:
                            attempt += 1
                            retry_of = span.span_id
                            continue
                        raise

            return wrapper

        return decorator

    def tool(self, name: Optional[str] = None, retries: int = 0) -> Callable:
        """Shorthand for a TOOL-kind span."""
        return self.span(name, kind=SpanKind.TOOL, retries=retries)

    def llm_call(self, name: Optional[str] = None, model: str = "", provider: str = "") -> Callable:
        """
        Trace an LLM call. Token usage is auto-extracted from OpenAI- and
        Anthropic-style response objects; cost is estimated from the model name.
        """

        def decorator(fn: Callable) -> Callable:
            inner = self.span(name or fn.__name__, kind=SpanKind.LLM)(fn)

            def _record(args, kwargs, result):
                span = None
                run = ctx.current_run()
                if run and run.spans:
                    # the span we just closed is the latest LLM span with this name
                    for s in reversed(run.spans):
                        if s.kind == SpanKind.LLM and s.name == (name or fn.__name__):
                            span = s
                            break
                if span is None:
                    return
                meta = LLMMetadata(model=model, provider=provider)
                prompt = kwargs.get("prompt") or (args[0] if args else "")
                meta.prompt_preview = _preview(prompt)
                usage = getattr(result, "usage", None) or (result.get("usage") if isinstance(result, dict) else None)
                if usage is not None:
                    get = (lambda k: getattr(usage, k, None)) if not isinstance(usage, dict) else usage.get
                    meta.input_tokens = int(get("prompt_tokens") or get("input_tokens") or 0)
                    meta.output_tokens = int(get("completion_tokens") or get("output_tokens") or 0)
                rmodel = getattr(result, "model", None) or (result.get("model") if isinstance(result, dict) else None)
                if rmodel:
                    meta.model = str(rmodel)
                meta.response_preview = _preview(result)
                meta.cost_usd = estimate_cost_usd(meta.model, meta.input_tokens, meta.output_tokens, self.cost_table)
                span.llm = meta
                self._check_budget()

            if inspect.iscoroutinefunction(fn):

                @functools.wraps(fn)
                async def async_wrapper(*args, **kwargs):
                    result = await inner(*args, **kwargs)
                    _record(args, kwargs, result)
                    return result

                return async_wrapper

            @functools.wraps(fn)
            def wrapper(*args, **kwargs):
                result = inner(*args, **kwargs)
                _record(args, kwargs, result)
                return result

            return wrapper

        return decorator

    # ------------------------------------------------------------------ #

    def _check_budget(self) -> None:
        run = ctx.current_run()
        if run is None:
            return
        reason = run.over_budget()
        if reason is None:
            return
        if self.on_budget == "raise":
            raise BudgetExceeded(run, reason)
        if self.on_budget == "pause":
            run.status = SpanStatus.PAUSED
            run.error = reason
        else:
            print(f"[agentlens] WARNING: {reason}")


# ---------------------------------------------------------------------- #
# zero-config module-level decorators (console exporter)
# ---------------------------------------------------------------------- #

_default = AgentLens()


def trace(name: Optional[str] = None, **kw) -> Callable:
    return _default.trace(name, **kw)


def tool(name: Optional[str] = None, **kw) -> Callable:
    return _default.tool(name, **kw)

File: sdk-ts/src/tracer.ts
/**
 * The tracer. Python uses decorators; JavaScript's are still awkward and
 * don't apply to plain functions, so this wraps instead — which also works
 * on arrow functions, closures, and third-party callbacks.
 *
 *   const lens = new AgentLens({ endpoint: 'http://localhost:7430' });
 *
 *   const search = lens.tool('web_search', async (q: string) => { … });
 *
 *   const agent = lens.trace('research_agent', async (q: string) => {
 *     return summarize(await search(q));
 *   }, { maxCostUsd: 0.1 });
 *
 * Sync and async functions both work; the return type is preserved.
 */

import { currentContext, currentRun, currentSpan, withContext, withSpan } from './context.js';
import { estimateCostUsd, extractUsage, type CostTable } from './cost.js';
import { ConsoleExporter, HttpExporter, type Exporter } from './exporters.js';
import { AgentRun, Span, preview, type SpanKind } from './models.js';
import { runEndEvent, runStartEvent, spanEvent } from './streaming.js';

export class BudgetExceededError extends Error {
  constructor(readonly run: AgentRun, readonly reason: string) {
    super(`AgentLens budget guard: ${reason} (run=${run.name})`);
    this.name = 'BudgetExceededError';
  }
}

export interface AgentLensOptions {
  endpoint?: string;
  apiKey?: string;
  exporter?: Exporter;
  costTable?: CostTable;
  onBudget?: 'raise' | 'pause' | 'warn';
  stream?: boolean;
}

export interface TraceOptions {
  tags?: string[];
  maxTotalTokens?: number;
  maxCostUsd?: number;
  metadata?: Record<string, unknown>;
}

export interface SpanOptions {
  kind?: SpanKind;
  retries?: number;
}

export interface LLMOptions {
  model?: string;
  provider?: string;
}

type AnyFn<A extends unknown[], R> = (...args: A) => R;

function errorText(e: unknown): string {
  if (e instanceof Error) return e.stack || `${e.name}: ${e.message}`;
  return String(e);
}

function isPromise(value: unknown): value is Promise<unknown> {
  return typeof (value as { then?: unknown } | null)?.then === 'function';
}

export class AgentLens {
  readonly exporter: Exporter;
  readonly costTable: CostTable;
  private readonly onBudget: 'raise' | 'pause' | 'warn';

  constructor(options: AgentLensOptions = {}) {
    this.exporter =
      options.exporter ??
      (options.endpoint
        ? new HttpExporter(options.endpoint, { apiKey: options.apiKey, stream: options.stream })
        : new ConsoleExporter());
    this.costTable = options.costTable ?? {};
    this.onBudget = options.onBudget ?? 'raise';
  }

  private emit(event: Record<string, unknown>): void {
    try {
      this.exporter.exportEvent?.(event);
    } catch {
      /* a live view is never worth failing a run over */
    }
  }

  /** Wrap a function as an agent entrypoint: it owns a run. */
  trace<A extends unknown[], R>(name: string, fn: AnyFn<A, R>, options: TraceOptions = {}): AnyFn<A, R> {
    return ((...args: A): R => {
      const run = new AgentRun(name, options.tags ?? []);
      run.maxTotalTokens = options.maxTotalTokens ?? null;
      run.maxCostUsd = options.maxCostUsd ?? null;
      run.metadata = options.metadata ?? {};

      const root = new Span(name, 'agent');
      root.inputs = preview(args);
      run.spans.push(root);

      const settle = (result: unknown, error: unknown): void => {
        if (error === undefined) {
          root.outputs = preview(result);
          root.finish('success');
          run.finish('success');
        } else if (error instanceof BudgetExceededError) {
          root.finish('paused', error.message);
          run.finish('paused', error.message);
        } else {
          root.finish('error', errorText(error));
          run.finish('error', error instanceof Error ? error.message : String(error));
        }
        this.emit(spanEvent(run, root, 'span_end'));
        this.emit(runEndEvent(run));
        try {
          this.exporter.export(run);
        } catch {
          /* tracing must not take the agent down */
        }
      };

      return withContext({ run, span: root }, () => {
        this.emit(runStartEvent(run));
        this.emit(spanEvent(run, root, 'span_start'));
        try {
          const result = fn(...args);
          if (isPromise(result)) {
            return result.then(
              (value) => {
                settle(value, undefined);
                return value;
              },
              (err) => {
                settle(undefined, err);
                throw err;
              },
            ) as R;
          }
          settle(result, undefined);
          return result;
        } catch (err) {
          settle(undefined, err);
          throw err;
        }
      });
    }) as AnyFn<A, R>;
  }

  /** Wrap a function as a child step of the current run. */
  span<A extends unknown[], R>(name: string, fn: AnyFn<A, R>, options: SpanOptions = {}): AnyFn<A, R> {
    const kind = options.kind ?? 'custom';
    const retries = options.retries ?? 0;

    return ((...args: A): R => {
      const ctx = currentContext();
      if (!ctx) return fn(...args); // untraced call passes straight through

      const attempt = (n: number, retryOf: string | null): R => {
        const span = new Span(name, kind, ctx.span?.span_id ?? null);
        span.retry_of = retryOf;
        span.inputs = preview(args);
        ctx.run.spans.push(span);
        this.emit(spanEvent(ctx.run, span, 'span_start'));

        const closeOk = (value: unknown): void => {
          span.outputs = preview(value);
          span.finish('success');
          this.emit(spanEvent(ctx.run, span, 'span_end'));
          this.checkBudget(ctx.run);
        };
        const closeErr = (err: unknown): void => {
          span.finish('error', errorText(err));
          this.emit(spanEvent(ctx.run, span, 'span_end'));
        };

        return withSpan(span, () => {
          try {
            const result = fn(...args);
            if (isPromise(result)) {
              return result.then(
                (value) => {
                  closeOk(value);
                  return value;
                },
                (err) => {
                  if (err instanceof BudgetExceededError) {
                    span.finish('success');
                    throw err;
                  }
                  closeErr(err);
                  // failed attempts stay in the DAG, linked by retry lineage
                  if (n < retries) return attempt(n + 1, span.span_id);
                  throw err;
                },
              ) as R;
            }
            closeOk(result);
            return result;
          } catch (err) {
            if (err instanceof BudgetExceededError) {
              span.finish('success');
              throw err;
            }
            closeErr(err);
            if (n < retries) return attempt(n + 1, span.span_id);
            throw err;
          }
        });
      };

      return attempt(0, null);
    }) as AnyFn<A, R>;
  }

  /** Shorthand for a tool-kind span. */
  tool<A extends unknown[], R>(name: string, fn: AnyFn<A, R>, options: { retries?: number } = {}): AnyFn<A, R> {
    return this.span(name, fn, { kind: 'tool', retries: options.retries ?? 0 });
  }

  /**
   * Trace an LLM call, reading token usage off the response and estimating
   * cost from the model name.
   */
  llmCall<A extends unknown[], R>(name: string, fn: AnyFn<A, R>, options: LLMOptions = {}): AnyFn<A, R> {
    const inner = this.span(name, fn, { kind: 'llm' });

    const record = (args: A, result: unknown): void => {
      const run = currentRun();
      if (!run) return;
      const span = [...run.spans].reverse().find((s) => s.kind === 'llm' && s.name === name);
      if (!span) return;

      const usage = extractUsage(result);
      const model = usage.model || options.model || '';
      span.llm = {
        model,
        provider: options.provider ?? '',
        input_tokens: usage.input,
        output_tokens: usage.output,
        total_tokens: usage.input + usage.output,
        cost_usd: estimateCostUsd(model, usage.input, usage.output, this.costTable),
        prompt_preview: preview(args[0]),
        response_preview: preview(result),
        temperature: null,
      };
      this.checkBudget(run);
    };

    return ((...args: A): R => {
      const result = inner(...args);
      if (isPromise(result)) {
        return result.then((value) => {
          record(args, value);
          return value;
        }) as R;
      }
      record(args, result);
      return result;
    }) as AnyFn<A, R>;
  }

  private checkBudget(run: AgentRun): void {
    const reason = run.overBudget();
    if (!reason) return;
    if (this.onBudget === 'raise') throw new BudgetExceededError(run, reason);
    if (this.onBudget === 'pause') {
      run.status = 'paused';
      run.error = reason;
    } else {
      console.warn(`[agentlens] WARNING: ${reason}`);
    }
  }
}

/** Attach a score to the active run. A no-op outside one, never a throw. */
export function score(
  name: string,
  value: number,
  options: { source?: string; threshold?: number; comment?: string; onSpan?: boolean } = {},
): void {
  const run = currentRun();
  if (!run) return;
  const threshold = options.threshold ?? null;
  run.scores.push({
    name,
    value,
    source: options.source ?? 'custom',
    threshold,
    passed: threshold === null ? null : value >= threshold,
    comment: options.comment ?? '',
    span_id: options.onSpan ? (currentSpan()?.span_id ?? null) : null,
    recorded_at: Date.now() / 1000,
  });
}
